src/mcfret.c

```diff
@@ -0,0 +1,621 @@
+    /* Looping over samples: Each sample represents a different starting point on the Hamiltonian trajectory */
+    for (samples=non->begin;samples<non->end;samples++){ 
+        ti=samples*non->sample;
+        if (non->cluster!=-1){
+            if (read_cluster(non,ti,&cl,Cfile)!=1){
+	              printf("Cluster trajectory file to short, could not fill buffer!!!\n");
+	              printf("ITIME %d\n",ti);
+	              exit(1);
+            }
+            /* Configuration belong to cluster */ 
+            if (non->cluster==cl){
+	              Ncl++;
+            }
+        }
+        if (non->cluster==-1 || non->cluster==cl){
+      /* Initialize time-evolution operator */
+            if (emission==1){
+                /* Read Hamiltonian */
+                read_Hamiltonian(non,Hamil_i_e,H_traj,ti);
+	          
+                /* Remove couplings between segments */
+                multi_projection_Hamiltonian(Hamil_i_e,non);
+
+                /* Use the thermal equilibrium as initial state */
+                density_matrix(vecr,Hamil_i_e,non,segments);
+		if (samples==0) write_matrix_to_file("Density.dat",vecr,non->singles);
+            } else { 
+                unitmat(vecr,non->singles);
+		if (samples==0) write_matrix_to_file("Unit.dat",vecr,non->singles);
+            }
+            clearvec(veci,non->singles*non->singles);
+        
+            /* Loop over delay */ 
+            for (t1=0;t1<non->tmax1;t1++){
+	        tj=ti+t1;
+	        /* Read Hamiltonian */
+                read_Hamiltonian(non,Hamil_i_e,H_traj,tj);
+	          
+                /* Remove couplings between segments */
+                multi_projection_Hamiltonian(Hamil_i_e,non);
+                
+                /* Update the MCFRET Response */
+                mcfret_response_function_sub(re_S_1, im_S_1,t1,non,vecr,veci);        
+                if (emission==0){         
+                   propagate_matrix(non,Hamil_i_e,vecr,veci,-1,samples,t1*x);
+                } else {
+		   propagate_matrix(non,Hamil_i_e,vecr,veci,1,samples,t1*x);
+		}	   
+            }/* We are closing the loop over time delays - t1 times */
+        } /* We are closing the cluster loop */
+
+         /* Update NISE log file */ 
+        log=fopen("NISE.log","a");
+        fprintf(log,"Finished sample %d\n",samples);
+          
+        time_now=log_time(time_now,log);
+        fclose(log);
+    }/* Closing the loop over samples */
+    
+    /* The calculation is finished, lets write output */
+    log=fopen("NISE.log","a");
+    fprintf(log,"Finished Calculating Response!\n");
+    fprintf(log,"Writing to file!\n");  
+    fclose(log);
+
+    if (non->cluster!=-1){
+        printf("Of %d samples %d belonged to cluster %d.\n",samples,Ncl,non->cluster);
+        if (samples==0){ /* Avoid dividing by zero */ 
+            samples=1;
+        }
+    }
+
+    fclose(H_traj);
+    if (non->cluster!=-1){
+        fclose(Cfile);
+    }
+
+    /* Normalize response */
```

Comment:
I would put normalization before writing any output to improve the function's code structure.

Reply:
Moved normalisation up a bit.